syntax = "proto3";

package private_join_and_compute;

import "applications/logistic_regression/gradient_descent_dp_messages.proto";
import "applications/secure_sigmoid/secure_sigmoid.proto";
import "applications/secure_sigmoid_new_mic/secure_sigmoid_new_mic.proto";

// Messages for "No DP" gradient descent

message StartMessage {}
message EndMessage {}

// Client acts as Party 0, and initiates the execution.
message GradientDescentDpClientMessage {
  oneof client_message_oneof {
		// These messages are to be executed before the gradient descent
		// loop
		MaskedXTransposeMessage client_masked_x_transpose_message = 1;
    
		// Sent at the start of each loop iteration
		StartMessage start_message = 15;
		
		// These messages constitute the 4-round sigmoid implementation.
    applications.RoundOneSigmoidMessage
		    client_sigmoid_round_1_message = 2;
		applications.RoundTwoSigmoidMessage
				client_sigmoid_round_2_message = 3;
		applications.RoundThreeSigmoidMessage
				client_sigmoid_round_3_message = 4;
		applications.RoundFourSigmoidMessage
				client_sigmoid_round_4_message = 5;
	  
		// These messages constitute the 6-round sigmoid implementation using the
		// new MIC gate.
    applications.RoundOneSigmoidNewMicMessage
		    client_sigmoid_new_mic_round_1_message = 9;
		applications.RoundTwoSigmoidNewMicMessage
				client_sigmoid_new_mic_round_2_message = 10;
		applications.RoundThreeSigmoidNewMicMessage
				client_sigmoid_new_mic_round_3_message = 11;
		applications.RoundThreePointFiveSigmoidNewMicMessage
			client_sigmoid_new_mic_round_3_point_5_message = 12;
		applications.RoundFourSigmoidNewMicMessage
				client_sigmoid_new_mic_round_4_message = 13;
		applications.RoundFiveSigmoidNewMicMessage
				client_sigmoid_new_mic_round_5_message = 14;
	
		// These messages are the last two messages of the gradient descent loop
	  XTransposeDMessage
		    client_x_transpose_d_message = 6;
	  ReconstructGradientMessage
		    client_reconstruct_gradient_message = 7;
  }
}

// Server acts as Party 1 and responds to the client
// initiating the execution.
message GradientDescentDpServerMessage {
  oneof server_message_oneof {
		// These messages are to be executed before the gradient descent
		// loop
		MaskedXTransposeMessage server_masked_x_transpose_message = 1;
    
		// These messages constitute the 4-round sigmoid implementation using only
		// FSS Gates
    applications.RoundOneSigmoidMessage
		    server_sigmoid_round_1_message = 2;
		applications.RoundTwoSigmoidMessage
				server_sigmoid_round_2_message = 3;
		applications.RoundThreeSigmoidMessage
				server_sigmoid_round_3_message = 4;
		applications.RoundFourSigmoidMessage
				server_sigmoid_round_4_message = 5;
	  
		// These messages constitute the 6-round sigmoid implementation using the
		// new MIC gate.
    applications.RoundOneSigmoidNewMicMessage
		    server_sigmoid_new_mic_round_1_message = 9;
		applications.RoundTwoSigmoidNewMicMessage
				server_sigmoid_new_mic_round_2_message = 10;
		applications.RoundThreeSigmoidNewMicMessage
				server_sigmoid_new_mic_round_3_message = 11;
		applications.RoundThreePointFiveSigmoidNewMicMessage
			server_sigmoid_new_mic_round_3_point_5_message = 12;
		applications.RoundFourSigmoidNewMicMessage
				server_sigmoid_new_mic_round_4_message = 13;
		applications.RoundFiveSigmoidNewMicMessage
				server_sigmoid_new_mic_round_5_message = 14;
				
		// These messages are the last two messages of the gradient descent loop		
	  XTransposeDMessage
		    server_x_transpose_d_message = 6;
	  ReconstructGradientMessage
		    server_reconstruct_gradient_message = 7;
		
		// Sent at the end of each loop iteration.
		EndMessage end_message = 15;
  }
}

// gRPC interface for Gradient Descent.
service GradientDescentDpRpc {
  // Handles a single protocol round.
  rpc Handle(GradientDescentDpClientMessage) returns (GradientDescentDpServerMessage) {}
}